File: largest_value_in_each_tree_row.cpp
// Time Complexity : O(n)
// Space Complexity : O(logn)
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
//1. Traverse and first add value at each level. 
//2. Update with max value
//3. Return the result
class Solution {
public:
    vector<int> result;
    vector<int> largestValues(TreeNode* root) {
        if(root==NULL)
            return result;
        dfs(root, 0);
        return result;
    }
    
    void dfs(TreeNode* root, int level){
        //return
        if(root==NULL)
            return;
        //logic
        if(result.size()==level){
            result.push_back(root->val);
        }else{
            result[level] = max(result[level], root->val);
        }
        level=level+1;
        dfs(root->left, level);
        dfs(root->right, level);
    }
};

File: LargestValueInEachTreeRow.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    public class LargestValueInEachTreeRow
    {
        /// Time Complexity : O(N) 
        // Space Complexity :O(N) 
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No 
        public IList<int> LargestValues(TreeNode root)
        {
            List<int> result = new List<int>();
            if (root == null) return result;

            Queue<TreeNode> q = new Queue<TreeNode>();
            q.Enqueue(root);

            while (q.Count != 0)
            {

                int size = q.Count;
                int max = int.MinValue;
                for (int i = 0; i < size; i++)
                {
                    TreeNode curr = q.Dequeue();
                    max = Math.Max(max, curr.val);
                    if (curr.left != null) q.Enqueue(curr.left);
                    if (curr.right != null) q.Enqueue(curr.right);
                }
                result.Add(max);
            }

            return result;

        }
    }
}
